From 1aa82aa5b654a77b5bbfd79fa3fb63b96121a051 Mon Sep 17 00:00:00 2001
From: Alex Cruz <alex@dirtyunicorns.com>
Date: Thu, 15 Nov 2018 22:21:11 -0600
Subject: [PATCH 12/19] Fix NPE with void android.view.VelocityTracker.clear()

06-21 14:39:46.546 30816 30816 E AndroidRuntime: FATAL EXCEPTION: main
06-21 14:39:46.546 30816 30816 E AndroidRuntime: Process: com.android.settings, PID: 30816
06-21 14:39:46.546 30816 30816 E AndroidRuntime: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.view.VelocityTracker.clear()' on a null object reference
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.widget.ScrollView.onTouchEvent(ScrollView.java:692)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.View.dispatchTouchEvent(View.java:10023)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2626)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2307)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at com.android.internal.policy.DecorView.superDispatchTouchEvent(DecorView.java:416)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at com.android.internal.policy.PhoneWindow.superDispatchTouchEvent(PhoneWindow.java:1808)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.app.Activity.dispatchTouchEvent(Activity.java:3065)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.WindowCallbackWrapper.dispatchTouchEvent(WindowCallbackWrapper.java:53)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at com.android.internal.policy.DecorView.dispatchTouchEvent(DecorView.java:378)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.View.dispatchPointerEvent(View.java:10243)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl$ViewPostImeInputStage.processPointerEvent(ViewRootImpl.java:4444)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:4312)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:3859)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:3912)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:3878)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:4005)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:3886)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:4062)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:3859)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:3912)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:3878)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:3886)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:3859)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl.deliverInputEvent(ViewRootImpl.java:6253)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl.doProcessInputEvents(ViewRootImpl.java:6227)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl.enqueueInputEvent(ViewRootImpl.java:6188)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl$WindowInputEventReceiver.onInputEvent(ViewRootImpl.java:6356)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.InputEventReceiver.dispatchInputEvent(InputEventReceiver.java:185)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.InputEventReceiver.nativeConsumeBatchedInputEvents(Native Method)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.InputEventReceiver.consumeBatchedInputEvents(InputEventReceiver.java:176)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl.doConsumeBatchedInput(ViewRootImpl.java:6327)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.ViewRootImpl$ConsumeBatchedInputRunnable.run(ViewRootImpl.java:6379)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:874)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.Choreographer.doCallbacks(Choreographer.java:686)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.Choreographer.doFrame(Choreographer.java:615)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:860)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:751)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:95)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:154)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6208)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:891)
06-21 14:39:46.546 30816 30816 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:781)

Change-Id: I415fc2e85bcdbe694f1d3a1bd76d1462fc8db78b
Signed-off-by: Genkzsz11 <Genkzsz11@gmail.com>
---
 core/java/android/widget/ScrollView.java | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/core/java/android/widget/ScrollView.java b/core/java/android/widget/ScrollView.java
index 15cd17b20f..e7aae34b3c 100644
--- a/core/java/android/widget/ScrollView.java
+++ b/core/java/android/widget/ScrollView.java
@@ -828,7 +828,9 @@ public class ScrollView extends FrameLayout {
                     if (overScrollBy(0, deltaY, 0, mScrollY, 0, range, 0, mOverscrollDistance, true)
                             && !hasNestedScrollingParent()) {
                         // Break our velocity if we hit a scroll barrier.
-                        mVelocityTracker.clear();
+                        if (mVelocityTracker != null) {
+                            mVelocityTracker.clear();
+                        }
                     }
 
                     final int scrolledDeltaY = mScrollY - oldY;
-- 
2.34.1

